hushline/routes.py

```diff
@@ -118,6 +118,27 @@ def profile(username: str) -> Response | str:
             flash("🫥 User not found.")
             return redirect(url_for("index"))
 
+        # If the encrypted message is stored in the session, use it to populate the form
+        scope = "submit_message"
+        if (
+            f"{scope}:salt" in session
+            and f"{scope}:contact_method" in session
+            and f"{scope}:content" in session
+        ):
+            try:
+                form.contact_method.data = decrypt_field(
+                    session[f"{scope}:contact_method"], scope, session[f"{scope}:salt"]
+                )
+                form.content.data = decrypt_field(
+                    session[f"{scope}:content"], scope, session[f"{scope}:salt"]
```

Comment:
Let's make sure the csrf token is also being used in the key derivation. The salt & csrf_token combination should be an adequately unique & session-bound value.

Reply:
The csrf_token isn't possible here actually, not without refactoring how this works. It's because we have two separate routes, loading the profile page:

```py
@app.route("/to/<username>", methods=["GET"])
def profile(username: str) -> Response | str:
    form = MessageForm()
```

And submitting a message:

```py
@app.route("/to/<username>", methods=["POST"])
def submit_message(username: str) -> Response | str:
    form = MessageForm()
```

When the user submits a message with an invalid CAPTCHA, it adds stuff to the session and then redirects:

```py
captcha_answer = request.form.get("captcha_answer", "")
if not validate_captcha(captcha_answer):
    # Encrypt the message and store it in the session
    scope = "submit_message"
    salt = generate_salt()
    session[f"{scope}:contact_method"] = encrypt_field(
        form.contact_method.data, scope, salt
    )
    session[f"{scope}:content"] = encrypt_field(form.content.data, scope, salt)
    session[f"{scope}:salt"] = salt

    return redirect(url_for("profile", username=username))
```

When the user subsequently loads the `profile` route, it creates a new instance of `MessageForm()`, which has a new CSRF token that's different from the previous one. So basically, at the point in time where we're ready to decrypt the fields, we've lost access to the original CSRF token.

That said, I think just the fact that each time the CAPTCHA is entered incorrectly it generates a new 128-bit salt that's used just for that scope should be fine for this use-case.

Alternatively, we could give anonymous users some sort of random session ID when the session cookie is first set and include that in the KDF, but it will also need to be stored in the session cookie, and I don't see how that's better than the unique salt.

Reply:
@micahflee Ok, I see. That makes sense. Ideally there would be some kind of binding information to a particular session. The thing is, using the same key for everything, especially in a multi-user setting, requires that there be very careful and clear domain separation. Another approach may in fact be to actually have distinct keys for each purpose. Though, that may require more thought and potentially structural changes. Key derivation should be fine for now if done carefully.

Apart from that, but what I also think would be sufficient in this case, is to increase the salt size to 32-bytes. The reasoning is that the key in question is a static 32 byte value. When the salt is 16 bytes, there's a max of 2<sup>128</sup> possible different keys that can be output from the key derivation. The number should be closer to 2<sup>256</sup>, otherwise much of the useful entropy from the key is inaccessible. Pseudo-randomly selecting a key from that larger space also decreases the likelihood of a repeat key to comfortable levels (a probability of 2<sup>-85.33...</sup> after about 2<sup>85.33...</sup> encryptions).